src/mono/mono/utils/mono-time.c

```diff
@@ -229,3 +235,84 @@ mono_100ns_datetime_from_timeval (struct timeval tv)
 
 #endif
 
+#if defined(HOST_DARWIN) || defined(HOST_LINUX)
+
+void
+mono_clock_init (void *clk_id)
+{
+#ifdef HOST_DARWIN
+	kern_return_t ret;
+
+	do {
+		ret = host_get_clock_service (mach_host_self (), SYSTEM_CLOCK, (clock_serv_t *)clk_id);
+	} while (ret == KERN_ABORTED);
+
+	if (ret != KERN_SUCCESS)
+		g_error ("%s: host_get_clock_service () returned %d", __func__, ret);
+#endif	
+}
+
+void
+mono_clock_cleanup (void *clk_id)
+{
+#ifdef HOST_DARWIN
+	kern_return_t ret;
+
+	do {
+		ret = mach_port_deallocate (mach_task_self (), *(clock_serv_t *)clk_id);
+	} while (ret == KERN_ABORTED);
+
+	if (ret != KERN_SUCCESS)
+		g_error ("%s: mach_port_deallocate () returned %d", __func__, ret);
+#endif
+}
+
+guint64
+mono_clock_get_time_ns (void *clk_id)
+{
+#ifdef HOST_DARWIN
+	
+	kern_return_t ret;
+	mach_timespec_t mach_ts;
+
+	do {
+		ret = clock_get_time (*(clock_serv_t *)clk_id, &mach_ts);
+	} while (ret == KERN_ABORTED);
+
+	if (ret != KERN_SUCCESS)
+		g_error ("%s: clock_get_time () returned %d", __func__, ret);
+
+	return ((guint64) mach_ts.tv_sec * 1000000000) + (guint64) mach_ts.tv_nsec;
+	
+#else
+	
+	struct timespec ts;
+
+	if (clock_gettime (*(clockid_t *)clock, &ts) == -1)
```

Comment:
Unless I'm missing something, this is passing the function pointer for `clock` cast to `clockid_t`, and should cause a crash on Linux. It's a good example of why the `void *` conversions scare me :p

Reply:
Oops, it is a typo. It should be `clk_id`, instead of `clock`.